### PR TITLE
Wasmi enclave

### PR DESCRIPTION
**Pull request by assafmo**

### CODE REVIEW

File: cosmwasm/lib/wasmi-runtime/Enclave_private.pem [hunk 1]
```diff
@@ -0,0 +1,39 @@
+-----BEGIN RSA PRIVATE KEY-----
```
Comment (toml01) [inline cosmwasm/lib/wasmi-runtime/Enclave_private.pem]: Whoopsy doopsy

File: cosmwasm/lib/Makefile [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+######## EDL Settings ########
+
+Enclave_EDL_Files := ./wasmi-runtime/Enclave_t.c ./wasmi-runtime/Enclave_t.h ./sgx-vm/Enclave_u.c ./sgx-vm/Enclave_u.h
+
+######## APP Settings ########
+
+App_Rust_Flags := --release
+App_SRC_Files := $(shell find sgx-vm/ -type f -name '*.rs') $(shell find sgx-vm/ -type f -name 'Cargo.toml')
+App_Include_Paths := -I ./sgx-vm -I./include -I$(SGX_SDK)/include -I$(CUSTOM_EDL_PATH)
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: Isn't the space after the fist `-I` not supposed to be there?

File: cosmwasm/lib/Makefile [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+.PHONY: all
+all: enclave-ffi-types $(App_Name) $(Signed_RustEnclave_Name)
+
+.PHONY: enclave-ffi-types
+enclave-ffi-types:
+	cd enclave-ffi-types && cargo build
+	mkdir -p ./sgx-vm/target/headers
+	mkdir -p ./wasmi-runtime/target/headers
+	cp -f ./enclave-ffi-types/target/headers/enclave-ffi-types.h ./wasmi-runtime/target/headers
+	cp -f ./enclave-ffi-types/target/headers/enclave-ffi-types.h ./sgx-vm/target/headers
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: All of this already happens as part of `enclave-ffi-types`'s build script, and you don't need to do it at all :) Any package that depends on `enclave-ffi-types` will have these files available when building, before it gets built itself

File: cosmwasm/lib/Makefile [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	Trts_Library_Name := sgx_trts
+	Service_Library_Name := sgx_tservice
+endif
+Crypto_Library_Name := sgx_tcrypto
+KeyExchange_Library_Name := sgx_tkey_exchange
+ProtectedFs_Library_Name := sgx_tprotected_fs
+
+RustEnclave_C_Files := $(wildcard ./wasmi-runtime/*.c)
+RustEnclave_C_Objects := $(RustEnclave_C_Files:.c=.o)
+RustEnclave_Include_Paths := -I$(CUSTOM_COMMON_PATH)/inc -I$(CUSTOM_EDL_PATH) -I$(SGX_SDK)/include -I$(SGX_SDK)/include/tlibc -I$(SGX_SDK)/include/stlport -I$(SGX_SDK)/include/epid -I ./wasmi-runtime -I./include
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: Near the end of the line, you probably have a redundant space in `-I ./wasmi-runtime`

File: cosmwasm/lib/wasmi-runtime/Enclave_private.pem (hunk 1 above)
Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem]:
```suggestion
-----BEGIN RSA PRIVATE KEY-----asdasd
```

Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem]: This is fine, it's just a test file from intel for now (google its content)

Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem]: Taken from here: https://github.com/intel/intel-sgx-ssl/blob/master/Linux/sgx/test_app/enclave/TestEnclave_private.pem

File: cosmwasm/lib/wasmi-runtime/Cargo.toml [hunk 5]
```diff
@@ -5,7 +5,43 @@ authors = ["Reuven Podmazo <reuven@enigma.co>"]
...
+sgx_tprotected_fs = { path = "../incubator-teaclave-sgx-sdk/sgx_tprotected_fs" }
+sgx_trts = { path = "../incubator-teaclave-sgx-sdk/sgx_trts" }
+sgx_tse = { path = "../incubator-teaclave-sgx-sdk/sgx_tse" }
+sgx_tseal = { path = "../incubator-teaclave-sgx-sdk/sgx_tseal" }
+sgx_tstd = { path = "../incubator-teaclave-sgx-sdk/sgx_tstd" }
+sgx_tunittest = { path = "../incubator-teaclave-sgx-sdk/sgx_tunittest" }
+sgx_types = { path = "../incubator-teaclave-sgx-sdk/sgx_types" }
+sgx_ucrypto = { path = "../incubator-teaclave-sgx-sdk/sgx_ucrypto" }
+sgx_unwind = { path = "../incubator-teaclave-sgx-sdk/sgx_unwind" }
+sgx_urts = { path = "../incubator-teaclave-sgx-sdk/sgx_urts" }
```
Comment (reuvenpo) [inline cosmwasm/lib/wasmi-runtime/Cargo.toml]: 1. The first section there says you're only pulling these dependencies when NOT building for SGX.
2. You then override the github links with local paths.
3. You should be pulling only the dependencies you need, and you're not using most of these
4. All of these are available on crates.io. please just specify normal crates.io dependencies.

File: cosmwasm/lib/.gitignore [hunk 6]
```diff
@@ -0,0 +1,9 @@
+lib*.a
+Cargo.lock
+Enclave_t.c
+Enclave_t.h
+Enclave_t.o
+
+Enclave_u.c
+Enclave_u.h
+Enclave_u.o
```
Comment (reuvenpo) [inline cosmwasm/lib/.gitignore]: Please use more specific .gitignore files in each directory. not each directory has all these files, and `wasmi-runtime` should definitely keep its `Cargo.lock` file.

File: cosmwasm/lib/Makefile [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+$(App_Enclave_u_Object): sgx-vm/Enclave_u.o
+	$(AR) rcsD $@ $^
+	cp $(App_Enclave_u_Object) ./lib
+
+$(App_Name): enclave-ffi-types $(App_Enclave_u_Object) $(App_SRC_Files)
+	@cd sgx-vm && SGX_SDK=$(SGX_SDK) cargo build $(App_Rust_Flags)
+	@echo "Cargo  =>  $@"
+	mkdir -p bin
+	cp $(App_Rust_Path)/sgx-vm ./bin
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: The rules that run cargo here should run as their own rule before `$(App_Enclave_u_Object)`, and instead of the `enclave-ffi-types` rule

File: cosmwasm/lib/wasmi-runtime/Enclave.config.xml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+<!-- Please refer to User's Guide for the explanation of each field -->
+<EnclaveConfiguration>
+  <ProdID>0</ProdID>
```
Comment (reuvenpo) [inline cosmwasm/lib/wasmi-runtime/Enclave.config.xml]: in enigma-core the ProdID was 1. you should check what it means

File: cosmwasm/lib/wasmi-runtime/Enclave_private.pem (hunk 1 above)
Comment (reuvenpo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem]: If it's not doing anything we should remove this file from the PR

File: cosmwasm/lib/wasmi-runtime/Xargo.toml [hunk 9]
```diff
@@ -0,0 +1,91 @@
+[dependencies]
+alloc = {}
+
+[dependencies.sgx_types]
+path = "../incubator-teaclave-sgx-sdk/sgx_types"
+stage = 1
```
Comment (reuvenpo) [inline cosmwasm/lib/wasmi-runtime/Xargo.toml]: Similarly to my comment above, this is n=probably unneeded.
also, you can rewrite this block as the following line, under the `[dependencies]` section:
```
sgx_types = { path = "../incubator-teaclave-sgx-sdk/sgx_types", stage = 1 }
```
and the other blocks below similarly

File: cosmwasm/lib/Makefile [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+wasmi-runtime/Enclave_t.o: $(Enclave_EDL_Files)
+	@$(CC) $(RustEnclave_Compile_Flags) -c wasmi-runtime/Enclave_t.c -o $@
+	@echo "CC   <=  $<"
+
+$(RustEnclave_Name): enclave compiler-rt wasmi-runtime/Enclave_t.o
+	cp incubator-teaclave-sgx-sdk/compiler-rt/libcompiler-rt-patch.a ./lib
+	@$(CXX) wasmi-runtime/Enclave_t.o -o $@ $(RustEnclave_Link_Flags)
+	@echo "LINK =>  $@"
+
+$(Signed_RustEnclave_Name): enclave-ffi-types $(RustEnclave_Name)
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: I believe that you don't need the `enclave-ffi-types` rule here at all, since the `$(RustEnclave_Name)` rule runs tthe `enclave` rule, which runs cargo/xargo before running anything else, and that should place the headers in their right place.

File: cosmwasm/lib/Makefile (hunk 3 above)
Comment (assafmo) [reply cosmwasm/lib/Makefile]: Both packages depend on `enclave-ffi-types` and they don't have these files, so...

Comment (assafmo) [reply cosmwasm/lib/Makefile]: It doesn't matter

File: cosmwasm/lib/Makefile (hunk 2 above)
Comment (assafmo) [reply cosmwasm/lib/Makefile]: It doesn't matter

File: cosmwasm/lib/Makefile (hunk 4 above)
Comment (assafmo) [reply cosmwasm/lib/Makefile]: It doesn't matter

File: cosmwasm/lib/wasmi-runtime/Cargo.toml (hunk 5 above)
Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Cargo.toml]: We only copied this from the sample project, trying to make it work rn

File: cosmwasm/lib/wasmi-runtime/Enclave_private.pem (hunk 1 above)
Comment (toml01) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem]: @assafmo I know, jk :)

File: cosmwasm/lib/Makefile (hunk 10 above)
Comment (assafmo) [reply cosmwasm/lib/Makefile]: Yep but we get an error about missing `enclave-ffi-types.h` in those directories so we added this rule for now

Comment (reuvenpo) [reply cosmwasm/lib/Makefile]: can you please show me the error when building only this rule?

File: cosmwasm/lib/Makefile [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+sgx-vm/Enclave_u.o: $(Enclave_EDL_Files)
+	@$(CC) $(App_C_Flags) -c sgx-vm/Enclave_u.c -o $@
+	@echo "CC   <=  $<"
+
+$(App_Enclave_u_Object): sgx-vm/Enclave_u.o
+	$(AR) rcsD $@ $^
+	cp $(App_Enclave_u_Object) ./lib
+
+foo:
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: Please rename this :smile:

File: cosmwasm/lib/Makefile (hunk 10 above)
Comment (reuvenpo) [reply cosmwasm/lib/Makefile]: resolved in call

File: cosmwasm/lib/wasmi-runtime/Cargo.toml (hunk 5 above)
Comment (reuvenpo) [reply cosmwasm/lib/wasmi-runtime/Cargo.toml]: resolved in call

File: cosmwasm/lib/Makefile (hunk 3 above)
Comment (reuvenpo) [reply cosmwasm/lib/Makefile]: resolved in call

File: cosmwasm/lib/Makefile [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+wasmi-runtime/Enclave_t.o: $(Enclave_EDL_Files)
+	@$(CC) $(RustEnclave_Compile_Flags) -c wasmi-runtime/Enclave_t.c -o $@
+	@echo "CC   <=  $<"
+
+$(RustEnclave_Name): enclave compiler-rt wasmi-runtime/Enclave_t.o
+	cp incubator-teaclave-sgx-sdk/compiler-rt/libcompiler-rt-patch.a ./lib
+	@$(CXX) wasmi-runtime/Enclave_t.o -o $@ $(RustEnclave_Link_Flags)
+	@echo "LINK =>  $@"
+
+$(Signed_RustEnclave_Name): enclave-ffi-types $(RustEnclave_Name)
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: remove the reference to `enclave-ffi-types`

File: cosmwasm/lib/Makefile [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+.PHONY: compiler-rt
+compiler-rt:
+	$(MAKE) -C incubator-teaclave-sgx-sdk/compiler-rt/ 2> /dev/null
+
+.PHONY: clean
+clean:
+	@rm -f $(App_Name) $(RustEnclave_Name) $(Signed_RustEnclave_Name) wasmi-runtime/*_t.* sgx-vm/*_u.* lib/*.a
+	@cd wasmi-runtime && cargo clean && rm -f Cargo.lock
+	@cd sgx-vm && cargo clean && rm -f Cargo.lock
+	@cd enclave-ffi-types && cargo clean && rm -f Cargo.lock
```
Comment (reuvenpo) [inline cosmwasm/lib/Makefile]: remove this line

File: go.mod [hunk 14]
```diff
@@ -5,6 +5,7 @@ go 1.13
 require (
 	github.com/cosmos/cosmos-sdk v0.38.1
 	github.com/gorilla/mux v1.7.3
+	github.com/pkg/errors v0.9.1
```
Comment (reuvenpo) [inline go.mod]: why did you add this?

File: cosmwasm/lib/wasmi-runtime/Xargo.toml (hunk 9 above)
Comment (reuvenpo) [reply cosmwasm/lib/wasmi-runtime/Xargo.toml]: done

File: cosmwasm/lib/wasmi-runtime/Enclave.config.xml (hunk 8 above)
Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave.config.xml:3]: Screenshot from 2020-04-08 00-40-51
Screenshot from 2020-04-08 00-40-59

Source: [Intel_SGX_Developer_Reference_Linux_2.6_Open_Source.pdf](https://github.com/enigmampc/EnigmaBlockchain/files/4447128/Intel_SGX_Developer_Reference_Linux_2.6_Open_Source.pdf)

Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave.config.xml:3]: Downloaded the pdf from https://download.01.org/intel-sgx/linux-2.6/docs/Intel_SGX_Developer_Reference_Linux_2.6_Open_Source.pdf

File: cosmwasm/lib/wasmi-runtime/Enclave_private.pem (hunk 1 above)
Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem:1]: It's used to sign the enclave so we do need this file: https://github.com/enigmampc/EnigmaBlockchain/blob/6cd448d376da812de940b7f20c870ad64f52c398/cosmwasm/lib/Makefile#L149

Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave_private.pem:1]: But we'll want to use our own private key in production. This is fine for now.

File: go.mod (hunk 14 above)
Comment (assafmo) [reply go.mod:8]: `go mod` discovered it in here:   https://github.com/enigmampc/EnigmaBlockchain/blob/6cd448d376da812de940b7f20c870ad64f52c398/cosmwasm/lib/incubator-teaclave-sgx-sdk/samplecode/ue-ra/ue-ra-client-go/cert.go#L10

So my IDE auto-added it to `go.mod`

File: cosmwasm/lib/wasmi-runtime/Enclave.config.xml (hunk 8 above)
Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave.config.xml:3]: It's our product number... We also have an ISV number from Intel we should put in this file.

Comment (assafmo) [reply cosmwasm/lib/wasmi-runtime/Enclave.config.xml:3]: In core it was 1 and in km it was 2 https://github.com/enigmampc/enigma-core/blob/3c04c2625741daa693cdc1a867db03bc1dc3ead2/enigma-principal/enclave/Enclave.config.xml#L3

It's just our product number. We also have an ISV number from Intel that we should put there.
